man/tojs.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "TOJS" "1" "October 2013" "skratchdot.com" "tojs"
.
.SH "Description"
\fBtojs\fR lets you convert files between plain text, javascript strings, and document\.write() statements\.
.
.P
It\'s original purpose, was to read in an \.html file, and output it as a series of document\.write() statements\. This is useful when creating bookmarklets\. You can code your \.html page, then run \fBtojs\fR on it\. Now all your bookmarklet has to do is include a script tag that points to the file you created with \fBtojs\fR\.
.
.P
Since version 1\.2\.0, \fBtojs\fR was converted to a node\.js project, and now supports a few different input/output formats\. To see the new options, check out the \fBUsage\fR and \fBOptions\fR sections below\.
.
.SH "Usage"
.
.nf

tojs [options] <files>
.
.fi
.
.SH "Options"
.
.nf

\-h, \-\-help           output usage information
\-v, \-\-version        output the version number
\-i, \-\-input [type]   The type of input [plain|js|jsvar|docwrite]
\-o, \-\-output [type]  The type of input [plain|js|jsvar|docwrite]
\-n, \-\-name [name]    If output is "jsvar", this is the variable name
\-\-oneline            Force output to be on one line only
\-\-single\-quotes      Use single\-quotes instead of double\-quotes for output
\-\-no\-var             var keyword will not be included in output
\-\-no\-open            document\.open() will not be included in output
\-\-no\-close           document\.close() will not be included in output
.
.fi
.
.SH "Examples"
.
.nf

$ tojs one\.txt two\.txt three\.txt
$ tojs \-o "docwrite" file\.html > file\.html\.js
$ tojs \-i "docwrite" \-o "plain" file\.html\.js > file\.html
$ tojs \-o "jsvar" \-\-oneline \-\-name "myVariable" file\.txt
$ echo \-e "one\entwo\enthree" | tojs
.
.fi
.
.SH "Installation"
.
.nf

npm install \-g tojs
.
.fi
.
.SH "Links"
Project Page \fIhttp://skratchdot\.com/projects/tojs/\fR
.
.P
Github Page \fIhttps://github\.com/skratchdot/tojs/\fR
.
.SH "Version History"
.
.SS "v1\.2\.2 (Released October 3, 2013)"
.
.IP "\(bu" 4
fixed "main" declaration in package\.json so library can be used from node apps (patch submitted by ryanramage \fIhttps://github\.com/ryanramage\fR)
.
.IP "" 0
.
.SS "v1\.2\.1 (Released January 2, 2012)"
.
.IP "\(bu" 4
adding man page
.
.IP "" 0
.
.SS "v1\.2\.0 (Released December 27, 2012)"
.
.IP "\(bu" 4
converting to a node\.js project so installation instructions have changed
.
.IP "\(bu" 4
command line options have changed
.
.IP "\(bu" 4
now supports the following input/output types:
.
.IP "\(bu" 4
plain
.
.IP "\(bu" 4
js
.
.IP "\(bu" 4
jsvar
.
.IP "\(bu" 4
docwrite
.
.IP "" 0

.
.IP "\(bu" 4
can now pipe data from stdin or use a list of files as input (similar to the unix cat command)
.
.IP "" 0
.
.SS "v1\.1\.0 (Released November 5, 2012)"
.
.IP "\(bu" 4
fixing formatting of help text
.
.IP "\(bu" 4
adding 2 new command line options
.
.IP "\(bu" 4
\fB\-o\fR | \fB\-\-no\-open\fR \-\- don\'t print the document\.open() statement
.
.IP "\(bu" 4
\fB\-c\fR | \fB\-\-no\-close\fR \-\- don\'t print the document\.close() statement
.
.IP "" 0

.
.IP "" 0
.
.SS "v1\.0\.0 (Released November 4, 2012)"
.
.IP "\(bu" 4
Initial Release
.
.IP "" 0



lib/tojs.js
/*jslint plusplus: true, regexp: true */
/*globals module */
(function (global) {
	'use strict';

	var tojs,
		encode,
		decode,
		trim,
		contains,
		cleanLine,
		overrideOptions,
		stringify,
		nonPlainTypes = ['js', 'jsvar', 'docwrite'];

	/**
	 * A cross-browser trim function. Removes the whitespace from
	 * the start and the end of the string.
	 */
	trim = function (str) {
		return str.replace(/^\s+/, '').replace(/\s+$/, '');
	};

	/**
	 * Return true if arr contains obj, otherwise return false;
	 */
	contains = function (arr, obj) {
		var i;
		for (i = 0; i < arr.length; i++) {
			if (arr[i] === obj) {
				return true;
			}
		}
		return false;
	};

	/**
	 * Override defaultOptions by doing validation on expected paramters names and types
	 */
	overrideOptions = function (defaultOptions, newOptions) {
		if (typeof newOptions === 'object') {
			// defaultOptions.type
			if (contains(nonPlainTypes, newOptions.type)) {
				defaultOptions.type = newOptions.type;
			}
			// defaultOptions.name
			if (typeof newOptions.name === 'string') {
				newOptions.name = newOptions.name.replace(/'|"/g, '');
				if (newOptions.name.length > 0) {
					defaultOptions.name = newOptions.name;
				}
			}
			// defaultOptions.isOneLine
			if (typeof newOptions.isOneLine === 'boolean') {
				defaultOptions.isOneLine = newOptions.isOneLine;
			}
			// defaultOptions.isSingleQuotes
			if (typeof newOptions.isSingleQuotes === 'boolean') {
				defaultOptions.isSingleQuotes = newOptions.isSingleQuotes;
			}
			// defaultOptions.isVarKeyword
			if (typeof newOptions.isVarKeyword === 'boolean') {
				defaultOptions.isVarKeyword = newOptions.isVarKeyword;
			}
			// defaultOptions.isOpen
			if (typeof newOptions.isOpen === 'boolean') {
				defaultOptions.isOpen = newOptions.isOpen;
			}
			// defaultOptions.isOpen
			if (typeof newOptions.isClose === 'boolean') {
				defaultOptions.isClose = newOptions.isClose;
			}
		}
		return defaultOptions;
	};

	/**
	 * A wrapper for JSON.stringify that allows single-quoted JSON strings
	 */
	stringify = function (str, isSingleQuotes) {
		str = JSON.stringify(str);
		if (isSingleQuotes) {
			str = str.replace(/^"/, '\'').replace(/"$/, '\'');
		}
		return str;
	};

	/**
	 * This attempts to clean a "javascript string" so it can
	 * be parsed.
	 * 
	 * It is expected to throw when passed a string it can't handle.
	 */
	cleanLine = function (str) {
		var docWriteStart = /^document\.write\(\s*/,
			docWriteEnd = /\s*\)$/,
			singleQuotesStart = /^\s*'/,
			singleQuotesEnd = /'\s*$/;

		// lines starting with var
		str = str.replace(/^\s*var\s+/, '');

		// lines starting with a variable name
		str = str.replace(/^\s*[^'"]+\s*=\s*/, '');

		// lines ending with a plus
		str = str.replace(/\s*\+\s*$/, '');

		// lines ending with a semicolon
		str = str.replace(/\s*\;\s*$/, '');

		// document.write() lines
		if (docWriteStart.test(str) && docWriteEnd.test(str)) {
			str = str.replace(docWriteStart, '');
			str = str.replace(docWriteEnd, '');
		}

		// single-quoted lines
		if (singleQuotesStart.test(str) && singleQuotesEnd.test(str)) {
			str = str.replace(singleQuotesStart, '');
			str = str.replace(singleQuotesEnd, '');
			str = '"' + str + '"';
		}

		// this might throw an error if we are dealing with an unexpected string type
		str = JSON.parse(str);

		return str;
	};

	/**
	 * tojs.encode
	 */
	encode = function (str, opts) {
		var i = 0,
			eol = ';',
			result = '',
			isLastLine = false,
			line = '',
			lines = [],
			options = {
				type : 'plain',
				name : 'temp',
				isOneLine : false,
				isSingleQuotes : false,
				isVarKeyword : true,
				isOpen : true,
				isClose : true
			};

		// override default options
		options = overrideOptions(options, opts);

		// return 'plain' results
		if (!contains(nonPlainTypes, options.type)) {
			return str;
		}

		// document.open
		if (options.type === 'docwrite' && options.isOpen) {
			result = 'document.open();' + (options.isOneLine ? '' : '\n');
		}

		// variable name
		if (options.type === 'jsvar') {
			result = (options.isVarKeyword ? 'var ' : '') + options.name + ' = ';
		}

		// build results
		if (options.isOneLine) {
			lines = [str];
		} else {
			lines = str.split('\n');
		}
		for (i = 0; i < lines.length; i++) {
			isLastLine = (i >= lines.length - 1);
			line = stringify(lines[i] + (isLastLine ? '' : '\n'), options.isSingleQuotes);
			if (options.type === 'docwrite') {
				line = 'document.write(' + line + ')';
			}
			// override eol
			if (!isLastLine && !options.isOneLine &&
					(options.type === 'js' || options.type === 'jsvar')) {
				eol = ' +';
			} else {
				eol = ';';
			}
			result += line + eol + (options.isOneLine ? '' : '\n');
		}

		// document.close
		if (options.type === 'docwrite' && options.isClose) {
			result = result + (options.isOneLine ? '' : '\n') + 'document.close();';
		}

		return result;
	};

	/**
	 * tojs.decode
	 */
	decode = function (str) {
		var i = 0,
			result = '',
			lines = [],
			open = 'document.open();',
			close = 'document.close();';

		// trim input string
		result = trim(str);

		// remove document.open
		if (result.indexOf(open) === 0) {
			result = result.slice(open.length);
		}

		// remove document.close
		if (result.indexOf(close) === result.length - close.length) {
			result = result.slice(0, result.length - close.length);
		}

		// split string into lines
		lines = result.split('\n');

		// clean each line. if the cleanLine() call fails, return the original string
		for (i = 0; i < lines.length; i++) {
			try {
				lines[i] = cleanLine(lines[i]);
			} catch (e) {
				return str;
			}
		}

		return lines.join('\n');
	};

	// setup tojs
	tojs = encode;
	tojs.encode = encode;
	tojs.decode = decode;

	// export via exports or store in our global object
	if (module && module.exports) {
		module.exports = tojs;
	} else {
		global.tojs = tojs;
	}
}(this));